app/models/bookable_slot.rb

```diff
@@ -27,61 +27,65 @@ def self.find_available_slot(start_at)
       .within_date_range(from, to)
       .select("#{quoted_table_name}.start_at::date as start_date")
       .select("#{quoted_table_name}.start_at")
-      .order('start_date asc')
+      .order('start_date asc, start_at asc')
       .group_by(&:start_date)
-      .transform_values { |value| value.map(&:start_at).uniq.sort }
+      .transform_values { |value| value.map(&:start_at).uniq }
   end
 
-  def self.without_appointments # rubocop:disable Metrics/MethodLength
+  def self.without_appointments(from = nil, to = nil) # rubocop:disable Metrics/MethodLength
     joins(<<-SQL
             LEFT JOIN appointments ON
               appointments.guider_id = #{quoted_table_name}.guider_id
+              #{reduce_by_range(:appointments, from, to)}
               AND NOT appointments.status IN (
                 #{Appointment.statuses['cancelled_by_customer']},
                 #{Appointment.statuses['cancelled_by_pension_wise']}
               )
               AND (
-                (appointments.start_at = #{quoted_table_name}.start_at AND appointments.end_at = #{quoted_table_name}.end_at)
-                OR
-                (appointments.start_at BETWEEN #{quoted_table_name}.start_at AND #{quoted_table_name}.end_at)
-                OR
-                (appointments.end_at BETWEEN #{quoted_table_name}.start_at AND #{quoted_table_name}.end_at)
+                appointments.start_at, appointments.end_at
+              ) OVERLAPS (
+                #{quoted_table_name}.start_at, #{quoted_table_name}.end_at
               )
             SQL
          )
       .where('appointments.start_at IS NULL')
   end
 
-  def self.without_holidays # rubocop:disable Metrics/MethodLength
+  def self.without_holidays(from = nil, to = nil) # rubocop:disable Metrics/MethodLength
     joins(<<-SQL
           LEFT JOIN holidays ON
             -- The holiday is specifically for the user, or it is for everyone
-            (holidays.user_id = #{quoted_table_name}.guider_id OR holidays.user_id IS NULL) AND
-            (
-              (holidays.start_at <= #{quoted_table_name}.start_at AND holidays.end_at >= #{quoted_table_name}.end_at)
-              OR
-              (holidays.start_at >= #{quoted_table_name}.start_at AND holidays.start_at < #{quoted_table_name}.end_at)
-              OR
-              (holidays.end_at > #{quoted_table_name}.start_at AND holidays.end_at <= #{quoted_table_name}.end_at)
+            (holidays.user_id = #{quoted_table_name}.guider_id OR holidays.user_id IS NULL)
+            #{reduce_by_range(:holidays, from, to)}
+            AND (
+              holidays.start_at, holidays.end_at
+            ) OVERLAPS (
+              #{quoted_table_name}.start_at, #{quoted_table_name}.end_at
             )
             SQL
          )
       .where('holidays.start_at IS NULL')
   end
 
+  def self.reduce_by_range(table, from, to)
+    return '' unless from && to
+
+    sanitize_sql(["AND (#{table}.start_at > ? AND #{table}.end_at < ?)", from, to])
```

Comment:
This clause is included only when a range is provided further up the stack. It's not to be used like a scope or arel.